### PR TITLE
Python Future Support

### LINKED ISSUE #1778: python: rpc interface design
As @SteVwonder mentioned in #1776 the python rpc interface could maybe use some renaming or redesign

> The naming could be changed though. rpc_create -> rpc_send. I am not sure what rpc_send should be renamed to in that case though. Maybe it should exist?

I'm not a python user and I think there is already design underway to support `flux_future_t` better in the python bindings. However, @trws mentioned it would be good to open little interface questions like this on the python bindings as issues so I'm doing that here.

My opinion (not that it should count for that much) is that there should be a single `rpc` method on the flux handle object, which returns a future with `then` `get` and `ready`, etc methods. It would cool (though I don't know if python allows this) to be able to do

```python
import flux
f = flux.Flux()
f.rpc ("kvs.ping").then (lambda fut,arg: print ("ping: route=%s", fut.get().route))
```

a synchronous RPC would be something like
```python
rc = f.rpc ("kvs.ping").get()
```

### PR DESCRIPTION
**Pull request by SteVwonder**
Adds a `Future` class that inherits from `WrapperPimpl`, and makes `RPC` a subclass of `Future`.  `f.rpc_create()` and `f.rpc_send()` are now dropped in favor of  `f.rpc()` and `f.rpc().get()`, respectively.  The `Future` class wasn't designed to be directly instantiated, it is more of an abstract class currently.

This PR also switches the bindings over to the [new-style cffi callbacks](https://cffi.readthedocs.io/en/latest/using.html#extern-python-and-void-arguments).  They should be faster and avoid any issues with SELinux [[1]](https://cffi.readthedocs.io/en/latest/using.html#callbacks-old-style).  The unfortunate side-effect being that the default compiler on TOSS3 doesn't appreciate the auto-generated C code:
```
_core.c:2272:3: warning: missing initializer for field ‘reserved1’ of ‘struct _cffi_externpy_s’ [-Wmissing-field-initializers]
   { "continuation_callback", 0 };
   ^
_core.c:174:11: note: ‘reserved1’ declared here
     void *reserved1, *reserved2;
```
So I had to throw a `-Wno-error=missing-field-initializers` into the python bindings `CFLAGS` to get things working.

Design questions for the broader group (@trws especially):

- Does the class inhertiance structure look reasonable?  It feels kind of weird to me to have `RPC.InnerWrapper` inherit from `Future.InnerWrapper`, but it works :man_shrugging: It also kind of bothers me that the Future class should not be directly instantiated (in the current design) **and** that there is nothing in the code preventing it (e.g., some `ABCmeta` magic).  Feels like one or the other of those two limitations should go away.
- What to do when `.then` times out?  What happens in `C`?  I assume `ETIMEDOUT` is set before the reactor calls the callback, but I don't see any examples in flux-core of anything checking for `ETIMEDOUT` inside a callback.
- If `.wait_for` times out, do we want to raise an exception, if so, which one?  Just a plain ol' `EnvironmentError` with errno set to ETIMEDOUT?  Or do we want to go the way that [`socket`](https://docs.python.org/2/library/socket.html#socket.timeout) and [`requests`](http://docs.python-requests.org/en/master/api/#requests.Timeout) did, and have a Flux-specific `timeout` exception?    I think what we do here is influenced by the answer to the previous question.
- Is returning an [`ffi.buffer`](https://cffi.readthedocs.io/en/latest/ref.html#ffi-buffer-ffi-from-buffer) of the raw future payload the "right thing" to do as the default implemetion for `Future.get()`?  I don't expect many, if any python users to use this interface, but for those that do, I was hoping to do something sane.  AFAIU with an ffi buffer, users can easily get a `bytes` object or initialize a struct; which seems pretty useful.

Another step towards #1649
Closes #1778

### CODE REVIEW

**Comment by garlick** [comment]
> What to do when `.then` times out?  What happens in `C`?  I assume `ETIMEDOUT` is set before the reactor calls the callback, but I don't see any examples in flux-core of anything checking for `ETIMEDOUT` inside a callback.

Right, the future is fulfilled with an error (ETIMEDOUT), and the fact that it is fulfilled then triggers the callback.  I don't think we've used it much, although unless the timeout condition was going to be special-cased, the error handling code might not explicitly test for it, e.g. something like
```c
if (flux_future_get (f, NULL) < 0)
    log_err_exit ("thing");
```
would handle the error and just print "thing: Connection timed out".

**Comment by SteVwonder** [comment]
Thanks @garlick! That makes sense.  I like your suggestion of not special casing the timeout.  We could just let the `EnvironmentError` get raised after a `.get()` and have `errno` set to `ETIMEDOUT`.  Given that, I'm thinking we should also raise an `EnvironementError` on `.wait_for` timeouts (rather than some custom timeout exception) for consistency.

**Comment by trws** [comment]
This looks like a good improvement to me, especially getting the callbacks cleaned up.  As to the inheritence, it’s reasonable, but I think I’d prefer if we make it such that just the Future class has an internal PIMPL construction, and we can pass through the extra prefixes as part of the Future constructor.  That way we have one less inheritence chain to worry about.

As to being something that shouldn’t be instantiated, it’s a bit like some of the other classes we have in that it can really only be created correctly by having a handle passed into it, or through a factory function basically.  I’d at least make it so that if the constructor is invoked without a handle, or with None, it throws an exception.

For buffers, I think that’s a reasonable thing for us to use, but it feels leaky.  I’m not sure I even want to allow a user to do that with the easy function though.  Maybe have `.get()` return a bytes array and `.get_raw()` give out the buffer with a warning that the representation of raw isn’t guaranteed stable or something?  In the common case I think we really want users to specify a get that “cooks” the result for them, but having the escape hatch there to get at the base makes sense.  Actually, maybe even allowing the user to *pass in* a buffer to read into like get normally does would be a good way to help avoid having that done too much?

**Comment by SteVwonder** [comment]
Thanks @trws for the feedback.

> I think I’d prefer if we make it such that just the Future class has an internal PIMPL construction, and we can pass through the extra prefixes as part of the Future constructor. That way we have one less inheritence chain to worry about.

Makes a lot of sense.  Done!

> I’d at least make it so that if the constructor is invoked without a handle, or with None, it throws an exception.

:+1: `handle` is now a required positional argument of the Future constructor

> For buffers, I think that’s a reasonable thing for us to use, but it feels leaky.....

Your suggestions are good ones.  Unfortunately, after I started trying to test the `get` method, I realized there is no way to get a size on the `flux_future_get` result, so producing a `bytes` string won't work.  I suppose letting the user pass in a buffer would work, but the more I think about it, I don't think we really get a whole lot by exposing the raw bytes right now.  Once we have a compelling, real-world use-case, I suggest we revisit this and decide how to handle it then.  For now, I opted to just remove `get` from the Future class and instead let the subclasses handle implementing their own `get`.

---

I also threw in some more tests, some more methods (`is_ready` and `wait_for`), and some fixes for anonymous RPCs getting garbage collected before their callbacks run.   I believe this PR is ready for a re-review.

**Comment by trws** [comment]
That all sounds good, thanks @stevwonder!

Quoting Stephen Herbein (2019-02-20 15:57:47)

**Comment by garlick** [comment]
Should there be some minimal documentation included with this PR for futures, or is that a topic for later on?  An example or two of futures in action might be useful here, at minimum, maybe as test programs driven by sharness?

**Comment by SteVwonder** [comment]
Thanks @garlick! As we discussed face-to-face:

> Should there be some minimal documentation included with this PR for futures

I opened a new issue (#2036) to track this.

> An example or two of futures in action might be useful here, at minimum, maybe as test programs driven by sharness?

I think [t0012-futures.py](https://github.com/flux-framework/flux-core/pull/2023/files#diff-227988a26fff29d570f0b6e02cd00276) is what you are looking for.

> or rewrite some flux commands in python, like flux-ping, flux-event, or flux-logger

I like this idea! As @grondo pointed out, we should solve #1993 first.

File: src/bindings/python/flux/wrapper.py [hunk 1]
```diff
@@ -162,6 +162,12 @@ def build_argument_translation_list(self, fun_type):
     def __call__(self, calling_object, *args_in):
         calling_object.ffi.errno = 0
         caller = calling_object.handle
+        if self.add_handle and caller is None:
```
Comment (chu11) [inline src/bindings/python/flux/wrapper.py]: This change seems to be somewhat independent of this commit?  I assume some bug was hit and you added it?  Perhaps needs a comment in the commit or should be in a separate commit?

Comment (SteVwonder) [reply src/bindings/python/flux/wrapper.py]: Oh. Good call.  That was added when I had a `destroy` method for `Future`.  But that method proved too tricky to handle consistency (too many corner cases), so I removed the method and then squashed the commit.  This stuck around.  I'll pull it out into its own commit (along with the relevant unit test, if there is one, if not, I'll make one)

Comment (SteVwonder) [reply src/bindings/python/flux/wrapper.py]: @chu11: I pulled this out into it's own commit and added a unit test in `t/python/wrapper.py` in the same commit.

**Comment by SteVwonder** [comment]
Rebased and addressed @chu11's comment.

**Comment by chu11** [comment]
LGTM.  Dunno if @trws wanted to give it another look over.

**Comment by garlick** [comment]
I think @trws is traveling, so let's put this in and we can always tweak it later if need be.  Thanks!
